polynote-kernel/src/main/scala/polynote/messages/Messages.scala

```diff
@@ -104,16 +127,92 @@ object RunCell extends MessageCompanion[RunCell](3)
+    // if the other edit is entirely before this edit, just shift this edit by the length delta
+    case ContentEdit(otherPos, otherDeleteLength, otherContent) if otherPos + otherDeleteLength <= pos =>
+      copy(pos = pos + otherContent.length - otherDeleteLength)
+
+    // if the other edit is entirely after this edit, nothing to do
+    case ContentEdit(otherPos, _, _) if otherPos >= pos + deleteLength => this
+
+    // if the other edit affects the region that was replaced by this edit, then they conflict
+    // we want the minimal edit that replaces the entire conflict region
+    case ContentEdit(otherPos, otherDeleteLength, otherContent) =>
+
+      // start at the earlier position
+      val start = math.min(pos, otherPos)
+
+      // They deleted to this position
+      val theirTarget = otherPos + otherDeleteLength
+
+      // I would have deleted up to this position...
+      val myTarget = pos + deleteLength
+
+      // So I should delete this much extra
+      val deleteExtra = math.max(0, myTarget - theirTarget) + math.max(0, otherPos - pos)
+
+      val combinedContent = if (otherPos <= pos) otherContent + content else content + otherContent
+
+      ContentEdit(start, otherContent.length + deleteExtra, combinedContent)
+  }
 }
 
-sealed trait NotebookUpdate
+sealed trait NotebookUpdate extends Message {
+  def globalVersion: Int
+  def localVersion: Int
+  def notebook: ShortString
+
+  def withVersions(global: Int, local: Int): NotebookUpdate = this match {
+    case u @ UpdateCell(_, _, _, _, _) => u.copy(globalVersion = global, localVersion = local)
+    case i @ InsertCell(_, _, _, _, _) => i.copy(globalVersion = global, localVersion = local)
+    case d @ DeleteCell(_, _, _, _)    => d.copy(globalVersion = global, localVersion = local)
+    case u @ UpdateConfig(_, _, _, _)  => u.copy(globalVersion = global, localVersion = local)
+    case l @ SetCellLanguage(_, _, _, _, _) => l.copy(globalVersion = global, localVersion = local)
+  }
+
+  // transform this update so that it has the same effect when applied after the given update
+  def rebase(prev: NotebookUpdate): NotebookUpdate = (this, prev) match {
+    case (i@InsertCell(_, _, _, cell1, after1), InsertCell(_, _, _, cell2, after2)) if after1 == after2 =>
+      // we both tried to insert a cell after the same cell. Transform the first update so it inserts after the cell created by the second update.
+      i.copy(after = Some(cell2.id))
+
+    case (u@UpdateCell(_, _, _, id1, edits1), UpdateCell(_, _, _, id2, edits2)) if id1 == id2 =>
+      // we both tried to edit the same cell. Transform first edits so they apply to the document state as it exists after the second edits are already applied.
+      val rebasedEdits1 = edits1.map {
+        edit => edits2.foldLeft(edit)((e1, e2) => e1.rebase(e2))
+      }
+      u.copy(edits = ShortList(rebasedEdits1))
+
+    // all other cases should be independent (TODO: they're not yet, though)
+    case _ => this
+
+  }
+}
```

Comment:
Just curious - the code in `withVersions` and `rebase` seems implementation specific as we are doing different things based on the type of `NotebookUpdate`. I would expect these methods to be abstract (or maybe with a default impl) in the trait and then extended in the trait's children. 

On the other hand, it is nice to see the entire implementation in one place and as it's a sealed trait we'd need to add new children to this file anyways. Is it a common pattern to do this sort of thing for sealed traits?

Reply:
I think for a sealed trait it's largely immaterial whether you do a match in the base class or use an abstract method. I usually start with the former, because it lets you more easily compare the logic of the different cases, and then if that gets unmanageable or doesn't seem to have a benefit then move the logic to subclasses.